src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+
+    do
+        listener.InstrumentPublished <-
+            fun i l ->
+                if i.Meter = meter && i.Description = cacheId then
+                    l.EnableMeasurementEvents(i)
+
+        listener.SetMeasurementEventCallback<int64>(fun k v _ _ -> Interlocked.Add(readings.GetOrAdd(k.Name, ref 0L), v) |> ignore)
+        listener.Start()
+
+    member this.Dispose() = listener.Dispose()
+
+    member val CacheId = cacheId
+
+    static member val Meter = meter
+
+    member val RecentStats = "-" with get, set
+
+    member this.TryUpdateStats(clearCounts) =
+        let ratio =
+            try
+                float readings["hits"].Value
+                / float (readings["hits"].Value + readings["misses"].Value)
+                * 100.0
+            with _ ->
+                Double.NaN
+
+        let stats =
+            [
+                for name in readings.Keys do
+                    let v = readings[name].Value
+
+                    if v > 0 then
+                        $"{name}: {v}"
+            ]
+            |> String.concat ", "
+            |> sprintf "%s | hit ratio: %s %s" this.CacheId (if Double.IsNaN(ratio) then "-" else $"%.1f{ratio}%%")
+
+        if clearCounts then
+            for r in readings.Values do
+                Interlocked.Exchange(r, 0L) |> ignore
+
+        if stats <> this.RecentStats then
+            this.RecentStats <- stats
+            true
+        else
+            false
+
+    // TODO: Should return a Map, not a string
+    static member GetStats(cacheId) =
+        observedCaches[cacheId].TryUpdateStats(false) |> ignore
+        observedCaches[cacheId].RecentStats
+
+    static member GetStatsUpdateForAllCaches(clearCounts) =
+        [
+            for i in observedCaches.Values do
+                if i.TryUpdateStats(clearCounts) then
+                    i.RecentStats
+        ]
+        |> String.concat "\n"
+
+    static member AddInstrumentation(cacheId) =
+        if observedCaches.ContainsKey cacheId then
+            invalidArg "cacheId" $"cache with name {cacheId} already exists"
+
+        observedCaches[cacheId] <- new CacheMetrics(cacheId)
+
+    static member RemoveInstrumentation(cacheId) =
+        observedCaches[cacheId].Dispose()
+        observedCaches.TryRemove(cacheId) |> ignore
+
+type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    let overCapacity =
+        CacheMetrics.Meter.CreateCounter<int64>("over-capacity", "count", cacheId)
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > totalCapacity then
+                overCapacity.Add 1L
+
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < totalCapacity then
```

Comment:
This is not thread-safe, but we dont really care about off-by-a-few mistakes here I guess.

Reply:
Yes, not worth it.